File: leetcode/daily-temperatures.py
# 739. Daily Temperatures
# 🟠 Medium
#
# https://leetcode.com/problems/daily-temperatures/
#
# Tags: Array - Stack - Monotonic Stack

import timeit
from typing import List


# Use a monotonic non-increasing stack to store temperatures seen for
# which we haven't found a higher temperature yet. Iterate over the
# input checking the current temperature against the ones stored in the
# stack, we pop all the lower temperatures and save the difference
# between indexes in the result array.
#
# Time complexity: O(n) - We process each temperature a max of 2 times.
# Space complexity: O(n) - The monotonic stack could grow to the same
# size as the input.
#
# This code failed with TLE but it seems to be a bug with the time
# limit in leetcode.
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        # Use a monotonic stack to keep the warmest recorded temperatures.
        stack = []
        # Use an array of the same length as the input to store the results.
        res = [0] * len(temperatures)
        for i in range(len(temperatures)):
            while stack and temperatures[stack[-1]] < temperatures[i]:
                idx = stack.pop()
                # We found a warmer temperature, write the number of days we
                # waited.
                res[idx] = i - idx
            # There are no lower temperatures in the array now, add this one.
            stack.append(i)
        return res


def test():
    executors = [Solution]
    tests = [
        [[30], [0]],
        [[30, 60, 90], [1, 1, 0]],
        [[30, 40, 50, 60], [1, 1, 1, 0]],
        [[73, 74, 75, 71, 69, 72, 76, 73], [1, 1, 4, 2, 1, 1, 0, 0]],
        [
            [73, 74, 75, 71, 69, 72, 73, 73, 73, 76, 73, 73, 73, 72, 100],
            [1, 1, 7, 2, 1, 1, 3, 2, 1, 5, 4, 3, 2, 1, 0],
        ],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for col, t in enumerate(tests):
                sol = executor()
                result = sol.dailyTemperatures(t[0])
                exp = t[1]
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for"
                    + f" test {col} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/daily-temperatures.rs
// 739. Daily Temperatures
// 🟠 Medium
//
// https://leetcode.com/problems/daily-temperatures/
//
// Tags: Array - Stack - Monotonic Stack

struct Solution;
impl Solution {
    /// Use a monotonic decreasing stack with the indexes of temperatures that we have seen
    /// already. Iterate over the input, for each element check its value against the top of the
    /// stack, while the temperature of the current element is greater than the temperature of the
    /// element at the top of the stack, pop it and update the result with the difference between
    /// the current index and the popped element index, since this is the first temperature after
    /// that one that was higher.
    ///
    /// Time complexity: O(n) - We visit each element, for each we do amortized O(1) work, we may
    /// need to pop n elements from the stack for one element, but that would mean that we did not
    /// pop any elements for any other.
    /// Space complexity: O(n) - The stack may grow to size n.
    ///
    /// Runtime 23 ms Beats 100%
    /// Memory 4.36 MB Beats 65.52%
    pub fn daily_temperatures(temperatures: Vec<i32>) -> Vec<i32> {
        let n = temperatures.len();
        let mut stack: Vec<usize> = vec![];
        let mut res = vec![0; n];
        for i in 0..n {
            while let Some(&top) = stack.last() {
                if temperatures[i] > temperatures[top] {
                    stack.pop();
                    res[top] = (i - top) as i32;
                } else {
                    break;
                }
            }
            stack.push(i);
        }
        res
    }
}

// Tests.
fn main() {
    let tests = [
        (
            vec![73, 74, 75, 71, 69, 72, 76, 73],
            vec![1, 1, 4, 2, 1, 1, 0, 0],
        ),
        (vec![30, 40, 50, 60], vec![1, 1, 1, 0]),
        (vec![30, 60, 90], vec![1, 1, 0]),
    ];
    println!("\n\x1b[92m» Running {} tests...\x1b[0m", tests.len());
    let mut success = 0;
    for (i, t) in tests.iter().enumerate() {
        let res = Solution::daily_temperatures(t.0.clone());
        if res == t.1 {
            success += 1;
            println!("\x1b[92m✔\x1b[95m Test {} passed!\x1b[0m", i);
        } else {
            println!(
                "\x1b[31mx\x1b[95m Test {} failed expected: {:?} but got {:?}!!\x1b[0m",
                i, t.1, res
            );
        }
    }
    println!();
    if success == tests.len() {
        println!("\x1b[30;42m✔ All tests passed!\x1b[0m")
    } else if success == 0 {
        println!("\x1b[31mx \x1b[41;37mAll tests failed!\x1b[0m")
    } else {
        println!(
            "\x1b[31mx\x1b[95m {} tests failed!\x1b[0m",
            tests.len() - success
        )
    }
}
